specgen/specgen/hellocap/listener.go

```diff
@@ -0,0 +1,200 @@
+package hellocap
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"sync"
+)
+
+const tlsRecordHeaderLen = 5
+
+type contextKey int
+
+// contextHelloKey is the context key for retrieving the ClientHello.
+const contextHelloKey contextKey = 0
+
+// FromRequest returns the TLS ClientHello sent by the remote. This request must have come through a
+// handler attached to the Server type in this package.
+func FromRequest(req *http.Request) ([]byte, error) {
+	hello, ok := req.Context().Value(contextHelloKey).(capturedHello)
+	if !ok {
+		return nil, errors.New("no attached ClientHello")
+	}
+	return hello.hello, hello.err
+}
+
+// attachHello is used to put a captured ClientHello on a http.Request context. This is how the
+// ClientHello, captured at the TCP level, is made available to users of this library.
+func attachHello(req *http.Request, hello capturedHello) *http.Request {
+	ctx := context.WithValue(req.Context(), contextHelloKey, hello)
+	return req.Clone(ctx)
+}
+
+// onHello is a callback invoked when a ClientHello is captured. Only one of hello or err will be
+// non-nil. This callback is not invoked for incomplete ClientHellos. In other words, err is non-nil
+// only if what has been read off the connection could not possibly constitute a ClientHello,
+// regardless of further data from the connection.
+type onHello func(hello []byte, err error)
+
+// capturingConn is a TCP connection and implements net.Conn. If a TLS ClientHello is sent by the
+// peer, capturingConn will invoke the provided onHello callback. This is the basic building block
+// by which we build a hello-capturing server.
+type capturingConn struct {
+	// Wraps a TCP connection.
+	net.Conn
+
+	helloRead bool
+	helloBuf  *bytes.Buffer
+	helloLock sync.Mutex // protects fields in this block
+
+	onHello onHello
+	onClose func()
+}
+
+func newCapturingConn(wrapped net.Conn, onHello onHello, onClose func()) *capturingConn {
+	return &capturingConn{
+		wrapped,
+		false,
+		new(bytes.Buffer),
+		sync.Mutex{},
+		onHello,
+		onClose,
+	}
+}
+
+func (c *capturingConn) Read(b []byte) (n int, err error) {
+	n, err = c.Conn.Read(b)
```

Comment:
`c.buf.ReadFrom(c.Conn)` or `c.Conn.WriteTo(c.buf)` will save a copy.

Reply:
I would also experiment with `ReadAtLeast`. It could block on very small reads, but it'll block until the entire hello is sent, which seems actually desirable. Then the method just becomes

```
// assert len(b) > 5 or return ErrShortBuffer (but that'll never happen)

n, err := io.ReadAtLeast(b, c.Conn, tlsRecordHeaderLen)
if err != nil {
    return n, err
}

// parse client hello from b

return n, err
```

Reply:
Whoops, I made a dumb mistake in the above - you're obviously trying to capture the whole hello, not just the header record. It still works, though, since it's very unlikely (actually maybe 100% impossible) `Read` will be called with <512 (or is it 517?) bytes. If you see that happening, then you can still block for the whole clienthello with `io.CopyN` into a buf.

Reply:
Sorry, what's the goal here?  We're fine to use `b` during execution of `Read`.  We also don't want to artificially block `Read`.  In practice, that shouldn't cause issues if the client is sending TLS.  It may cause issues if the client does not send TLS (we'd have to implement the same logic `crypto/tls` uses to detect that the peer is not speaking TLS).

Reply:
The goal would be to remove the extra copy, simplify the code and remove the need for locking. You can just remove the extra copy with `ReadFrom` if you don't want to try to bite off the whole hello with `ReadAtLeast`, although I still think that could be nice. It can't actually block; it'll just error if there aren't enough bytes to read, which would mean a malfunctioning client or something.

Reply:
So I guess what you're proposing is something like:

1.  Use `io.ReadAtLeast` to only read the record header.
2. Parse the record length out of the header.
3. Use `io.ReadAtLeast` to read the full record.

There are some tricky pieces to this.  Most importantly, there is no guarantee about the size of `b`.  In theory, it may not even be large enough to hold the header.  In practice, it is certainly possible that `b` cannot hold the full record.  So you'll need to add logic to buffer the extra bytes when necessary.  Then, after you've read the full record and find that it's larger than `b`, what do you do with the remainder?  You'll need to hang onto those buffered bytes for subsequent reads.

Another challenge is that `io.ReadAtLeast` will in fact block until the requested number of bytes are read.  I'm not totally decided on whether that's an issue, but we'd need to think it through.  There may be additional challenges.  This is as far as I felt like going down this line of thinking.

Essentially what you're proposing is how `crypto/tls` reads records.  I'd encourage you to read through the library to get a sense for (a) the actual complexity of what you're proposing and (b) the operations already taking place in handling a TLS record.  You're trying to avoid copying ~1KB or less to save some cycles in an encrypted network connection.  Which is already copying (buffering) every byte.  In a library which explicitly does not care about performance.

Reply:
`readRecordOrCCS` is the function which kicks off this logic in `crypto/tls`, in case that's helpful.

Reply:
Concretely, here's what I am/was proposing:

```
func readClientHello(r io.Reader, b []byte) (int, error) {
	// This will return ErrShortBuffer if len(b) < tlsRecordLen. 5 bytes should
	// always be safe to read.
	off, err := io.ReadAtLeast(r, b, tlsRecordLen)
	if err != nil {
		return off, err
	}

	helloLen, err := lengthFromHeader(b[:tlsRecordLen])
	if err != nil {
		return off, err
	}

	// Again, this will return ErrShortBuffer if we don't have enough capacity
	// to read the whole hello. I would be extremely surprised to see that in
	// practice, since the stdlib tries very hard to avoid short reads. Also,
	// in reality, the whole client hello will almost always have been read in
	// first ReadAtLeast already.
	needed := helloLen - (off - tlsRecordLen)
	if needed > 0 {
		n, err := io.ReadAtLeast(r, b[off:], needed)
		if err != nil {
			return off + n, err
		}

                off += n
	}

	// ClientHello is in b[tlsRecordLen:off+1]

	return off, nil
}
```

(My original idea was to assume clienthellos would be smaller than 512 bytes, but doing usually-one-but-sometimes-two reads is also good.)

Even if we don't care about performance, I think it's clearer what's happening, and it doesn't require the reader or the code to track state across calls to `Read`. That said, if we really don't care about performance, then it's obviously moot, or at least academic.